### PR TITLE
🧑‍🔬 Prototype: Enforcing MFA for owners of top 100 downloaded gems (UI) (Option 2️⃣ )

### PR DESCRIPTION
**Pull request by bettymakes**
Contributes to https://github.com/Shopify/code-foundations/issues/166

## 🧪 What problem does this prototype solve?
**TL;DR – This prototype requires owners of the top 100 downloaded gems to enable MFA.** Once enabled, those users will not be able to disable MFA. They can however, deregister a device and register with a new one. However, they will not be able to access any other views on `rubygems.org` that require authentication until MFA is enabled. 

---

Account takeovers is second most common vulnerability that's exploited. In enforcing MFA, we add more friction and increase complexity of successfully exploiting that vulnerability.

In an ideal world, all accounts registered on `rubygems.org` require MFA. Recognizing that policy may require a lot more change management efforts for all stakeholders to be bought-in, we brainstormed options that'll get us from 1 to 2, as opposed to 1 to 10. 

And so, instead of enforcing MFA on _everyone_, an incremental approach could be to target the highest-risk accounts. We identify higher risk accounts as `accounts belonging to users who are owners of the top 100 most downloaded gems`, since the blast radius of a nefarious actor taking over high risk accounts is quite large. 

## 🧬 How does this prototype work?
This introduces 3 different distinct user flows:

🔹 **When a user is an owner of a top 100 gem, with MFA disabled**
Once the user signs in to `rubygems.org`, they are redirected to the MFA set up view. All views that require authentication will not be accessible to the user until MFA is set up. They will still be able to view any publicly accessible views that do not require authentication.

The option to disable MFA is removed for these users. However, if they want to switch their authentication device, they can deregister the current device, and register a new one. Once their current device is deregistered, they cannot access any views that require authentication until they register their new device.

⚠️ The risk here is that it's possible for a user to deregister their device, technically disabling MFA. And, if they do not follow through with the second step of registering a new device, their account is once again _more vulnerable_ to account takeover attacks.

🔹 **When a user is an owner of a top 100 gem, with MFA enabled**
This is the happy path! Once the user signs in to `ruybgems.org`, they can access all the views.

🔹 **When a user is not an owner of a top 100 gem**
There are no changes to this user's experience on `rubygems.org`, as they are not in the cohort this prototype is proposing to enforce MFA on.

## 🤷‍♀️ How does this approach differ from option 1️⃣ 
In option 1️⃣ , the top 100 gem owners are determined dynamically.

The approach to this implementation is to add a `mfa_required` attribute to `users`. The benefits of this approach is that the MFA enforcement is consistent. There may be gems that fluctuate along the threshold "top 100", going from `#100` to `#101` for instance and vice versa. In the case where a gem falls out of range, but was once within the top 100, we'll continue to enforce MFA on their accounts. However, with this approach, we'll likely need to introduce some mechanism (e.g. scheduled job) to update the list of user accounts that should have MFA enforced. 

## 📝 Other notes
* This prototype only focuses on the UI experience and not CLI.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -49,8 +49,8 @@ def redirect_to_signin
     redirect_to sign_in_path, alert: t("please_sign_in")
   end
 
-  def mfa_required?
-    current_user&.mfa_required?
+  def mfa_non_compliant?
```
Comment (bettymakes) [inline app/controllers/application_controller.rb:52]: I'm open to suggestions for naming this concept. Alternatives I thought of were:
- invalid_mfa_config
- invalid_mfa_configuration
- invalid_mfa_setting
- non_compliant_mfa
- mfa_non_compliant

File: script/mfa_required_for_user [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env ruby
```
Comment (bettymakes) [inline script/mfa_required_for_user:1]: Ashley and I initially worked on this as a script. However, I couldn't get it to run properly so I converted it over to a rake task.

File: lib/tasks/mfa_required_for_user.rake [hunk 3]
```diff
@@ -0,0 +1,19 @@
+namespace :mfa_required_for_user do
```
Comment (bettymakes) [inline lib/tasks/mfa_required_for_user.rake:1]: This is effectively the same as the script: `script/mfa_required_for_user` 
Since this was a POC, I didn't want to get too hung up on why it wasn't working as a script. After converting it to a rake task, I was able to execute it successfully. 

This also opens up a conversation that @aellispierce & I had about whether this should be a script or a rake task. Not sure if anyone knows of or has any strong heuristics around this.

**Comment by bettymakes** [comment]
Closing in favour of #10 since it's the amalgamation of all the prototypes.
